### ISSUE TITLE
User account self unlock

### ISSUE THREAD

**Issue by sztamas** [muccg/rdrf#503]
Allow users to reactivate their account by requesting a token to be sent to their email address and visiting it.

Original issue: muccg/rdrf-ccg#940

**Comment by sztamas** [comment]
This is a new feature that can be disabled/enabled by adjusting `settings.ACCOUNT_SELF_UNLOCK_ENABLED`. It is on by default.

When you have it enabled you get a new menu item in the main menu bar "Unlock Account", right next to "Password reset". The menu item only appears while you're not logged into the system.

The unlock account is very similar to password reset in many respects.

The user is expected to enter an email address. If the email address is registered in the system they will get an account unlock link. The link can be used only once and expires the same way the password reset link expires.
When they visit the link they will have to change their password which will also unlock their account.

An admin can decide to explicitly prevent a user to unlock their account by setting the new "Prevent self unlock" flag on the user edit page.
In that case the user won't receive any emails when they request to unlock their account.


There are several new templates that can be used to customise the pages/email messages. The following are analogous to the password reset templates:

- `registration/account_unlock_form.html`
- `registration/account_unlock_done.html`
- `registration/account_unlock_subject.txt`
- `registration/account_unlock_email.html`
- `registration/account_unlock_complete.html`

There is also a template that doesn't have a password reset counterpart:
  `registration/account_unlock_requested_but_account_not_locked_email.html`
This is an email that gets sent when someone is asking for an Account Unlock, but their account is NOT locked.

The password reset functionality has also been slightly altered.
Currently if a user is inactive (locked) they won't receive an email when they are requesting a Password Reset. This functionality is unchanged if the unlock account functionality is NOT enabled.

However, when the unlock account functionality is enabled (`settings.ACCOUNT_SELF_UNLOCK_ENABLED` is `True`) and the user isn't explicitly prevented to unlock their account, they will receive the password reset link email even if their account isn't active.

On visiting the link and changing their password the system will also unlock their account.

**Comment by knapier** [comment]
Looks great @sztamas . 

How do users know their account has been locked? Are they notified by email only, or will they be notified their account is locked when they attempt to log-in?

**Comment by sztamas** [comment]
There are some improvements in other tickets I did:
  1. The text has been changed to say that their account will be locked if they enter the password incorrectly 3 times (or whatever the settings is for the registry)
  2. If you fail to log in from the same IP more than 3 times in 60 minutes (configurable in settings again) the error message on the login page will change to "Your account has been locked"

Some notes regarding 2:
  - the error message will change to "Your account has been locked" even if your username doesn't exist in the system. That is to prevent giving out information about which emails are registered
  - it can sometimes give false positives, but I think those cases will be very limited so I would consider 2 an improvement overall
